PS-7526: Unexpected quoting and dropping of comments in DROP TABLE co…

nitendra-bhosle opened the pull request:
…mmands

https://jira.percona.com/browse/PS-7526

Analysis

Feature request to add binlog_ddl_skip_rewrite option
so that when it's set ON log drop table statements as it is for
single table drop statements without comments being striped out
and without adding quotes.
Generate error for multi table drop table statements

Fix

Added new option binlog_ddl_skip_rewrite , which is
by default OFF, and could be enabled by 'SET binlog_ddl_skip_rewrite = ON'
If enabled single table drop ddl statements will be logged as it is
without striping of comments and adding quotes.
Generate error for multi table drop table statements because
multi table drop statements are error prone if tables contain
both normal tables and temporary tables.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3372,7 +3385,17 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
       thd->thread_specific_used = true;
 
-      if (built_query.write_bin_log()) goto err_with_rollback;
+      if (thd->variables.binlog_ddl_query_log_events) {
+        if (thd->binlog_query(
+                THD::STMT_QUERY_TYPE, thd->query().str, thd->query().length,
+                drop_ctx.has_base_atomic_tables(), false /* direct */,
+                false /* suppress_use */, 0 /* errcode */)) {
+          goto err_with_rollback;
+        }
+
```
I think abort after engine commit is not good idea.  If you want to abort, better to do it at start of DDL? (or atleast after server decides that this is multi table DROP)

satya-bodapati replied:
8.0 has atomic DDL but only for InnoDB

nitendra-bhosle replied:
This behaviour is consistent with behaviour where binlog_ddl_query_log_events is 'off'.It's just that instead of built_query , we are using thd->query()

satya-bodapati replied:
Well, consistent with the way other errors are handled but not the intention. In other cases, there is actually a problem like unable to update DD, engine execution failure, etc. Here we upfront know that we will not support multi-table DROP.

Also, this is not exactly what we agreed upon.
1. No option
2. log query single table drop as is
3. Let multi table drop to proceed with existing re-write logic. ie. comments will be lost for multi table DROP.

Please check with customer if the above behavior is OK. Essentially it is a pick a poison scenario
1. Allow multi table DROP with loss of comments
2. Add new option and with option enabled, Disallow multi table DROP (throw error to client)

satya-bodapati replied:
If I consider the general use case, I feel a new option that enables "comments" and drops functionality is not right.

nitendra-bhosle replied:
For multi table drop statements , I am raising error before DDL start .
```
if (table_count > 1) {
      my_error(ER_DROP_MULTI_TABLE, MYF(0), "binlog_ddl_query_log_events");
      return true;
    }
```

nitendra-bhosle replied:
> 1. Allow multi table DROP with loss of comments
> 2. Add new option and with option enabled, Disallow multi table DROP (throw error to client)

Sure  , I will check with customer for this on JIRA ticket

nitendra-bhosle replied:
Customer feels option 2  "Add new option and with option enabled, Disallow multi table DROP (throw error to client)" is better.

satya-bodapati replied:
Right, proceeding with 2..

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,12 @@ static Sys_var_bool Sys_binlog_rows_query(
 
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_query_log_events",
+    "Write drop table statements having single table into binary log for",
+    SESSION_VAR(binlog_ddl_query_log_events), CMD_LINE(OPT_ARG), DEFAULT(false),
```
Also mention here this optino disableds multi-table table drop, ie DROP TABLE t1,t2.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,19 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+      ++table_count;
+      if (table_count > 1) break;
+    }
+
+    if (table_count > 1) {
+      my_error(ER_DROP_MULTI_TABLE, MYF(0), "binlog_ddl_query_log_events");
+      return true;
```
Please explain as comment why multi table drop is disallowed?

venkatesh-prasad-v replied:
+1

venkatesh-prasad-v commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,12 @@ static Sys_var_bool Sys_binlog_rows_query(
 
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_query_log_events",
+    "Write drop table statements having single table into binary log for",
```
Please complete the sentence

venkatesh-prasad-v commented on mysql-test/suite/binlog/t/binlog_stm_drop_tbl_comment.test:
```diff
@@ -0,0 +1,5 @@
+# This is a wrapper for binlog_drop_tbl_with_comment.test so that the same
+# test case can be used ror both statement and row based bin logs
```
```suggestion
# test case can be used for both statement and row based logging
```

venkatesh-prasad-v commented on mysql-test/suite/binlog/t/binlog_row_drop_tbl_comment.test:
```diff
@@ -0,0 +1,5 @@
+# This is a wrapper for binlog_drop_tbl_with_comment.test so that the same
+# test case can be used ror both statement and row based bin logs
```
```suggestion
# test case can be used for both statement and row based logging
```

venkatesh-prasad-v commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,19 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+    }
+
+    if (table_count > 1) {
+      my_error(ER_DROP_MULTI_TABLE, MYF(0), "binlog_ddl_query_log_events");
```
Please use `LogErr` instead of `my_error`.

nitendra-bhosle replied:
I can use , but it might be inconsistent with other drop table errors.

venkatesh-prasad-v replied:
Not an issue. Please ignore this comment.

venkatesh-prasad-v commented on mysql-test/extra/binlog_tests/binlog_drop_tbl_with_comment.test:
```diff
@@ -0,0 +1,36 @@
+
+DROP DATABASE test7526;
+
+source include/show_binlog_events.inc;
```
Can we use `--` for mysqltest commands?
```suggestion
--source include/show_binlog_events.inc
```

venkatesh-prasad-v reviewed (changes requested):
Please update the commit message to tell why we had to disallow multi-table updates.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,25 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+    }
+
+    /*
+      When 'binlog_ddl_query_log_events' option is enabled
+      logging query as is for multi table drop statements
+      are error prone if tables contain both normal tables
+      and temperary tables.Hence not supported
```
Error prone? Please explain the issues a bit more elaborately?

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,13 @@ static Sys_var_bool Sys_binlog_rows_query(
     ON_CHECK(check_session_admin));
 
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_query_log_events",
```
this option is not about logging DDL events to binlog, we already do that.
Can we use the name that the option does?
binlog_ddl_retain_comments?
binlog_ddl_disallow_query_rewrite?

Any others that u or @venkatesh-prasad-v can think of?

venkatesh-prasad-v replied:
As discussed over slack, I prefer naming it to `binlog_ddl_skip_rewrite`.

satya-bodapati replied:
Ok with it :+1:

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,13 @@ static Sys_var_bool Sys_binlog_rows_query(
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_query_log_events",
+    "Write drop table statements having single table into binary log,"
+    "Raise error for multi table drop statements",
```
Similarly comments are also telling something else...

venkatesh-prasad-v commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,13 @@ static Sys_var_bool Sys_binlog_rows_query(
     DEFAULT(false), NO_MUTEX_GUARD, NOT_IN_BINLOG,
     ON_CHECK(check_session_admin));
 
+static Sys_var_bool Sys_binlog_ddl_query(
```
Whenever we add a new server variable, we will need to rerecord a few main suite testcases like
```
1. all_persisted_variables
2. mysqld--help-notwin
and others
```

Please run the main suite once and check if there any test failures.

venkatesh-prasad-v replied:
Also, please update the recent Jenkins link as a comment or in the end of the description section.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,30 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+      does not not work as expected if tables contain both normal tables
+      and temperary tables,consider these two cases.
+      Case1: Statement's like 'drop table t1,t2' where t1 is normal table
+      but t2 is temperory table  will fail on slave while applying log.
```
temporary

satya-bodapati replied:
and t2 is a temporary table, will fail on the slave because temporary table will not be present on the slave.

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,13 @@ static Sys_var_bool Sys_binlog_rows_query(
 
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_skip_rewrite",
+    "Logs the input query  into the binary log without the server rewrite."
```
remove double space between query and into

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1607,6 +1607,13 @@ static Sys_var_bool Sys_binlog_rows_query(
+static Sys_var_bool Sys_binlog_ddl_query(
+    "binlog_ddl_skip_rewrite",
+    "Logs the input query  into the binary log without the server rewrite."
+    "This action is useful when the server-side DDL query strips comments.",
```
I would say, "Without server rewrite, DDL queries will be logged with comments"

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,30 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+      logging query as is for multi table drop statements
+      does not not work as expected if tables contain both normal tables
+      and temperary tables,consider these two cases.
+      Case1: Statement's like 'drop table t1,t2' where t1 is normal table
```
Statements*

satya-bodapati replied:
t1 is a

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,30 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+
+    /*
+      When 'binlog_ddl_skip_rewrite' option is enabled
+      logging query as is for multi table drop statements
```
logging query without rewrite

nitendra-bhosle commented:
Jenkins : https://ps80.cd.percona.com/job/percona-server-8.0-param/1010/

venkatesh-prasad-v commented on mysql-test/r/mysqld--help-notwin.result:
```diff
@@ -100,6 +100,10 @@ The following options may be given as the first argument:
  CRC32; default is CRC32.
+ --binlog-ddl-skip-rewrite 
+ Without server rewrite, DDL queries will be logged with
+ comments.This action is useful when the server-side DDL
```
Let there be space between `.` and `T`.
```suggestion
 comments. This action is useful when the server-side DDL
```

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -1609,7 +1609,7 @@ static Sys_var_bool Sys_binlog_rows_query(
     "binlog_ddl_skip_rewrite",
-    "Logs the input query  into the binary log without the server rewrite."
+    "Without server rewrite, DDL queries will be logged with comments."
     "This action is useful when the server-side DDL query strips comments.",
```
Since we mention about the comments, we can remove this line "This action is useful.." Sorry, I wasn't clear about this in last comment.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3095,14 +3095,14 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
-      Case2 : Statement's like 'DROP TABLE t1 / *!80024 ,t2 * /'
-      will generate single table or multi table drop statements
-      depending on the mysql version.
+      and temporary tables,consider these two cases.
```
re-arrange a bit to fit 80 chars on each line?

satya-bodapati reviewed (changes requested):
Almost there

satya-bodapati commented on mysql-test/extra/binlog_tests/binlog_drop_tbl_with_comment.test:
```diff
@@ -0,0 +1,36 @@
+
+DROP TABLE t,t1;
+
+DROP DATABASE test7526;
```
what about DROP DATABASE with comments? are these retained without your fix?

satya-bodapati commented on mysql-test/extra/binlog_tests/binlog_drop_tbl_with_comment.test:
```diff
@@ -0,0 +1,36 @@
+SET binlog_ddl_skip_rewrite = ON;
+
+/* comment1 */ CREATE TABLE t (a VARCHAR(100));
+/* comment1 */ DROP TABLE t;
```
can you please add one more scenario with a comment at last?

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,29 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+    }
+
+    /*
+      When 'binlog_ddl_skip_rewrite' option is enabled logging query without
```
```suggestion
      When 'binlog_ddl_skip_rewrite' option is enabled, logging query without
```

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,29 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+      tables and temporary tables,consider these two cases.
+      Case1: Statements like 'drop table t1,t2' where t1 is a normal table
+      and t2 is a temporary table, will fail on the slave because temporary
+      table will not be present on the slave. Case2 : Statements like 'DROP
```
Better to keep Case 2: on separate line

satya-bodapati reviewed (changes requested):
please add two more scenarios as requested
1. DROP TABLE with comments at end
2. DROP DATABASE with comments(with and without option enabled)

satya-bodapati reviewed (changes requested):
already left comments before you requested review :wink:

venkatesh-prasad-v commented on sql/sql_table.cc:
```diff
@@ -3086,6 +3086,29 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+    for (TABLE_LIST *table = tables; table; table = table->next_local) {
+      ++table_count;
+      if (table_count > 1) break;
+    }
```
I'm sorry for getting this a bit late. In my opinion, having this check inside `mysql_rm_table_no_locks()` can break the PXC cluster for multi-table drop statements.

1. Let's say we execute `/* comment1 */ DROP TABLE t1,t2;` on node1 of a PXC cluster.
2. The query is first replicated to other nodes via TOI, the the query `/* comment1 */ DROP TABLE t1,t2;` will be replicated to other nodes.
3.  On node1 the query reaches `mysql_rm_table_no_locks()`. Since it is a multi table drop, the query on node1 fails with `ER_DROP_MULTI_TABLE` error.
4. Whereas the same query on other nodes becomes successful since the applier threads on other nodes will have `thd->variables.binlog_ddl_skip_rewrite` set to `false`.
5. This causes the error voting protocol to kick out node1 from the cluster since the query was successful on other nodes.

For fixing this, I would suggest to move this check before we replicate in PXC. i.e, before https://github.com/percona/percona-xtradb-cluster/blob/8.0/sql/sql_parse.cc#L4115-L4124

venkatesh-prasad-v commented on sql/sql_table.cc:
```diff
@@ -3372,7 +3395,17 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
+                THD::STMT_QUERY_TYPE, thd->query().str, thd->query().length,
+                drop_ctx.has_base_atomic_tables(), false /* direct */,
+                false /* suppress_use */, 0 /* errcode */)) {
+          goto err_with_rollback;
```
In a replication scenario when binlog_ddl_skip_rewrite option is enabled on source(master) server, the query is written as-is to the binary log. But however, when log_slave_updates in enabled on the replica server, the applier thread rewrites the query in the replica's binlog omitting the comments.
i.e, on source server's binlog, we will have
```
| master-bin.000001 | 2205 | Query          |         1 |        2319 | use `test`; /* comment1 */ DROP TABLE t1 /* xid=41 */                    |
| master-bin.000001 | 2319 | Anonymous_Gtid |         1 |        2396 | SET @@SESSION.GTID_NEXT= 'ANONYMOUS'                                     |
| master-bin.000001 | 2396 | Query          |         1 |        2510 | use `test`; /* comment1 */ DROP TABLE t2 /* xid=42 */                    |
+-------------------+------+----------------+-----------+-------------+--------------------------------------------------------------------------+
```
whereas on replica's binlog, we will have
```
| slave-bin.000002 | 2338 | Query          |         1 |        2465 | use `test`; DROP TABLE `t1` /* generated by server */ /* xid=26 */       |
| slave-bin.000002 | 2465 | Anonymous_Gtid |         1 |        2549 | SET @@SESSION.GTID_NEXT= 'ANONYMOUS'                                     |
| slave-bin.000002 | 2549 | Query          |         1 |        2676 | use `test`; DROP TABLE `t2` /* generated by server */ /* xid=27 */       |
+------------------+------+----------------+-----------+-------------+--------------------------------------------------------------------------+
```

For fixing this, we must also include this check
```
(thd->system_thread == SYSTEM_THREAD_SLAVE_SQL) ||
(thd->system_thread == SYSTEM_THREAD_SLAVE_WORKER)
```
while checking for `thd->variables.binlog_ddl_skip_rewrite`, and we must allow the query to be re-written when it is a replication applier thread.

venkatesh-prasad-v commented:
Also, please update the commit message to reflect the new name of the server variable. As of now, it still has `binlog_ddl_query_log_events` in the description.

satya-bodapati commented on sql/sql_table.cc:
```diff
@@ -3374,7 +3374,8 @@ bool mysql_rm_table_no_locks(THD *thd, TABLE_LIST *tables, bool if_exists,
           thd->system_thread == SYSTEM_THREAD_SLAVE_SQL ||
-          thd->system_thread == SYSTEM_THREAD_SLAVE_WORKER) {
+          thd->system_thread == SYSTEM_THREAD_SLAVE_WORKER ||
+          thd->is_binlog_applier()) {
```
so independent of ddl_skip_rewrite option, as long it is binlog_applier, we log query as is (skip rewrite)?

venkatesh-prasad-v replied:
The rationale is that the query rewrite should be avoided for replication applier and binlog applier threads (during PITR), so that the query is logged as is. So, this should independent of binlog_ddl_skip_rewrite option.

satya-bodapati replied:
:+1:

satya-bodapati commented on sql/sql_parse.cc:
```diff
@@ -3622,6 +3622,30 @@ int mysql_execute_command(THD *thd, bool first_level) {
+        size_t table_count = 0;
+        for (TABLE_LIST *table = all_tables; table; table = table->next_local) {
+          ++table_count;
+          if (table_count > 1) break;
+        }
+
+        /*
+          When 'binlog_ddl_skip_rewrite' option is enabled, logging query
+          without rewrite does not not work as expected if tables contain
+          both normal tables and temporary tables,consider these two cases.
+          Case1: Statements like 'drop table t1,t2' where t1 is a normal
+          table and t2 is a temporary table, will fail on the slave because
+          temporary table will not be present on the slave.
+          Case2 : Statements like 'DROP TABLE t1 / *!80024 ,t2 * /' will
+          generate single table or multi table drop statements depending
+          on the mysql version.
+        */
+        if (table_count > 1) {
+          my_error(ER_DROP_MULTI_TABLE, MYF(0), "binlog_ddl_skip_rewrite");
+          goto error;
+        }
```
we can combine these two if statements to
if (table_count > 1) {
  my_error()
  goto error;
}

satya-bodapati commented on sql/sql_parse.cc:
```diff
@@ -3622,6 +3622,29 @@ int mysql_execute_command(THD *thd, bool first_level) {
+              Case1: Statements like 'drop table t1,t2' where t1 is a normal
+              table and t2 is a temporary table, will fail on the slave because
+              temporary table will not be present on the slave.
+              Case2 : Statements like 'DROP TABLE t1 / *!80024 ,t2 * /' will
```
Just for consistency reason, remove space between Case2 and :

satya-bodapati commented on sql/sys_vars.cc:
```diff
@@ -125,8 +125,8 @@
 #include "sql/sql_lex.h"
-#include "sql/sql_locale.h"     // my_locale_by_number
-#include "sql/sql_parse.h"      // killall_non_super_threads
+#include "sql/sql_locale.h"  // my_locale_by_number
+#include "sql/sql_parse.h"   // killall_non_super_threads
```
if done by clang format, leave it as is. If not, please revert it.

nitendra-bhosle replied:
> if done by clang format, leave it as is. If not, please revert it.

Done by clang format

satya-bodapati replied:
:+1:

satya-bodapati reviewed (changes requested):
As discussed on slack,
1. sys_var testsuite fix (add new testcase file)
2. other binlog persist failures
3. a file conflict detected by github has to be fixed

satya-bodapati commented:
@nitendra-bhosle also Jenkins run after all fixes

nitendra-bhosle commented:
New jenkins run : https://ps80.cd.percona.com/job/percona-server-8.0-param/1018/
https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1019/

nitendra-bhosle commented:
> As discussed on slack,
> 
> 1. sys_var testsuite fix (add new testcase file)
> 2. other binlog persist failures
> 3. a file conflict detected by github has to be fixed

@satya-bodapati  : Addressed these points

satya-bodapati commented on mysql-test/suite/sys_vars/t/binlog_ddl_skip_rewrite_basic.test:
```diff
@@ -0,0 +1 @@
+SELECT @@global.binlog_ddl_skip_rewrite;
```
Please add more scenarios. Check existing testcases for dynamic.
1. try session
2. set invalid values
3. set allowed values
4. try SET var= default (default is allowed value) and check variable value

venkatesh-prasad-v commented on mysql-test/suite/sys_vars/t/binlog_ddl_skip_rewrite_basic.test:
```diff
@@ -0,0 +1,152 @@
+SELECT @@session.binlog_ddl_skip_rewrite;
+
+########################################################################### 
+#       Change the value of sql_log_bin to invalid value                  #
```
Here and everywhere 

`s/sql_log_bin/binlog_ddl_skip_rewrite/`

satya-bodapati commented on mysql-test/suite/sys_vars/t/binlog_ddl_skip_rewrite_basic.test:
```diff
@@ -0,0 +1,152 @@
+#     Check if the value in GLOBAL Table contains variable value        #
+#########################################################################
+
+--disable_warnings
```
what warnings do we get? May be u can remove this and everywhere else in testcase

satya-bodapati commented on mysql-test/suite/sys_vars/t/binlog_ddl_skip_rewrite_basic.test:
```diff
@@ -0,0 +1,152 @@
+######################################################################## 
+#      change the value of binlog_ddl_skip_rewrite to a valid value    #
+######################################################################## 
+
+SET @@session.binlog_ddl_skip_rewrite = 0;
+SELECT @@session.binlog_ddl_skip_rewrite;
+SET @@session.binlog_ddl_skip_rewrite = 1;
+SELECT @@session.binlog_ddl_skip_rewrite;
+
```
add few scenarios for global as well. SET GLOBAL

satya-bodapati reviewed (changes requested):
Newly added test is still failing:  binlog.binlog_stm_drop_tbl_comment

210721 13:38:45 [ 54%] binlog.binlog_stm_drop_tbl_comment 'mix' w6  [ fail ]
        Test ended at 2021-07-21 13:38:45

CURRENT_TEST: binlog.binlog_stm_drop_tbl_comment
--- /tmp/results/PS/mysql-test/suite/binlog/r/binlog_stm_drop_tbl_comment.result	2021-07-21 14:26:58.000000000 +0300
+++ /tmp/results/PS/mysql-test/var/6/log/binlog_stm_drop_tbl_comment.reject	2021-07-21 16:38:45.236094938 +0300
@@ -30,6 +30,11 @@
 /* comment start */ DROP DATABASE test7526 /* comment end */;
 include/show_binlog_events.inc
 Log_name	Pos	Event_type	Server_id	End_log_pos	Info
+binlog.000001	#	Query	#	#	use `test`; CREATE TABLE t1 (a INT)
+binlog.000001	#	Query	#	#	BEGIN
+binlog.000001	#	Query	#	#	use `test`; INSERT INTO t1 VALUES(1)
+binlog.000001	#	Xid	#	#	COMMIT /* XID */
+binlog.000001	#	Query	#	#	use `test`; DROP TABLE `t1` /* generated by server */
 binlog.000001	#	Query	#	#	CREATE DATABASE test7526
 binlog.000001	#	Query	#	#	use `test7526`; /* comment1 */ CREATE TABLE t (a VARCHAR(100))
 binlog.000001	#	Query	#	#	use `test7526`; DROP TABLE `t` /* generated by server */

https://ps80.cd.percona.com/job/percona-server-8.0-pipeline/32020/consoleFull

@nitendra-bhosle can you please ensure clean Jenkins run before next review?

nitendra-bhosle commented:
> Newly added test is still failing: binlog.binlog_stm_drop_tbl_comment
> 
> 210721 13:38:45 [ 54%] binlog.binlog_stm_drop_tbl_comment 'mix' w6 [ fail ]
> Test ended at 2021-07-21 13:38:45
> 
> CURRENT_TEST: binlog.binlog_stm_drop_tbl_comment
> --- /tmp/results/PS/mysql-test/suite/binlog/r/binlog_stm_drop_tbl_comment.result 2021-07-21 14:26:58.000000000 +0300
> +++ /tmp/results/PS/mysql-test/var/6/log/binlog_stm_drop_tbl_comment.reject 2021-07-21 16:38:45.236094938 +0300
> @@ -30,6 +30,11 @@
> /* comment start _/ DROP DATABASE test7526 /_ comment end _/; include/show_binlog_events.inc Log_name Pos Event_type Server_id End_log_pos Info +binlog.000001 # Query # # use `test`; CREATE TABLE t1 (a INT) +binlog.000001 # Query # # BEGIN +binlog.000001 # Query # # use `test`; INSERT INTO t1 VALUES(1) +binlog.000001 # Xid # # COMMIT /_ XID _/ +binlog.000001 # Query # # use `test`; DROP TABLE `t1` /_ generated by server _/ binlog.000001 # Query # # CREATE DATABASE test7526 binlog.000001 # Query # # use `test7526`; /_ comment1 _/ CREATE TABLE t (a VARCHAR(100)) binlog.000001 # Query # # use `test7526`; DROP TABLE `t` /_ generated by server */
> 
> https://ps80.cd.percona.com/job/percona-server-8.0-pipeline/32020/consoleFull
> 
> @nitendra-bhosle can you please ensure clean Jenkins run before next review?

@satya-bodapati  : This test is failing because binlog.binlog_row_drop_tbl test is not doing proper cleaup 
binlog.binlog_row_drop_tbl binlog.binlog_row_drop_tbl_comment --no-reorder , fails.
I will update on PR after successful jenkins run.

satya-bodapati commented on mysql-test/suite/rpl_nogtid/t/rpl_row_mysqlbinlog.test:
```diff
@@ -144,7 +144,7 @@ connection master;
 select "--- Test 1 Load from Dump binlog file --" as "";
 
---exec $MYSQL -e "source $MYSQLTEST_VARDIR/tmp/master.sql"
+--exec $MYSQL -c -e "source $MYSQLTEST_VARDIR/tmp/master.sql"
```
are we fixing an unrelated test failure?

nitendra-bhosle replied:
No, this test case is related to thd->is_binlog_applier() check

satya-bodapati replied:
-c would send comments to server. Without this test change, how does it fail?

nitendra-bhosle replied:
Without the above mentioned condition check
DROP TABLE IF EXISTS t1,t2,t3 /* generated by server */
was getting logged

With the above mentioned condition check , we now log
DROP TABLE IF EXISTS `t1`,`t2`,`t3`

Because of this , position offsets of statements in binlog 
are different and hence test case was failing. And using -c was required to fix that.

satya-bodapati replied:
Does it mean,  customers/clients has to fix their client scripts to use "-c" ? :thinking:

satya-bodapati replied:
@nitendra-bhosle @venkatesh-prasad-v ^

venkatesh-prasad-v replied:
This change was required because the test does this

1. Do some transactions and record binlog position
2. Run mysqlbinlog and save the output
3. Clear the data and reset binary logs
4. Replay the binary log and verify the binlog positions

and the expectation of the test is that binlog position from step 1 should be same as step 4.

In step 1, the `DROP TABLE`  is recorded in the binlog as `DROP TABLE /* generated by server*/` But when the same binlog is replayed, it is logged as `DROP TABLE` without the comments.

So, in order to retain the comments during binlog replay, it is expected that clients should pass comments using `-c` option.

satya-bodapati replied:
Ok, thanks. So client doesn't need  to modify to use -c . (only the this particular test has to)

satya-bodapati reviewed (commented):
it looks good now. Do you have any Jenkins link (after all the test fixes)?

@venkatesh-prasad-v OK to squash?

nitendra-bhosle commented:
> it looks good now. Do you have any Jenkins link (after all the test fixes)?
> 
> @venkatesh-prasad-v OK to squash?

Latest jenkins run : https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1019/

venkatesh-prasad-v commented:
Looks good! Please go ahead and squash all commits.

satya-bodapati commented:
> > it looks good now. Do you have any Jenkins link (after all the test fixes)?
> > @venkatesh-prasad-v OK to squash?
> 
> Latest jenkins run : https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1019/

Still there are more failures that needs to be fixed. :cry: 
```
---
210722 19:15:41 [ 28%] rpl_nogtid.rpl_row_ignorable_event 'row' w6  [ fail ]
        Test ended at 2021-07-22 19:15:41

CURRENT_TEST: rpl_nogtid.rpl_row_ignorable_event
--- /tmp/results/PS/mysql-test/suite/rpl_nogtid/r/rpl_row_ignorable_event.result	2021-07-22 18:45:36.000000000 +0300
+++ /tmp/results/PS/mysql-test/var/6/log/rpl_row_ignorable_event.reject	2021-07-22 22:15:40.239966299 +0300
@@ -1062,7 +1062,7 @@
 master-bin.000001	#	Write_rows	#	#	table_id: # flags: STMT_END_F
 master-bin.000001	#	Xid	#	#	COMMIT /* XID */
 master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t1` /* generated by server */
-master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2` /* generated by server */
+master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2`
 # SHOW BINLOG EVENTS ON 2nd CONNECTION TO MASTE
 set @@session.binlog_rows_query_log_events= off;
 # Test statements from the 2nd connection have no comment event
@@ -1146,7 +1146,7 @@
 slave-bin.000001	#	Write_rows	#	#	table_id: # flags: STMT_END_F
 slave-bin.000001	#	Xid	#	#	COMMIT /* XID */
 slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t1` /* generated by server */
-slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2` /* generated by server */
+slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2`
---
```
Please fix all test failures. After ensuring all failures are existing failures, please re-request for review.

satya-bodapati commented:
> > > it looks good now. Do you have any Jenkins link (after all the test fixes)?
> > > @venkatesh-prasad-v OK to squash?
> > 
> > 
> > Latest jenkins run : https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1019/
> 
> Still there are more failures that needs to be fixed. cry
> 
> ```
> ---
> 210722 19:15:41 [ 28%] rpl_nogtid.rpl_row_ignorable_event 'row' w6  [ fail ]
>         Test ended at 2021-07-22 19:15:41
> 
> CURRENT_TEST: rpl_nogtid.rpl_row_ignorable_event
> --- /tmp/results/PS/mysql-test/suite/rpl_nogtid/r/rpl_row_ignorable_event.result	2021-07-22 18:45:36.000000000 +0300
> +++ /tmp/results/PS/mysql-test/var/6/log/rpl_row_ignorable_event.reject	2021-07-22 22:15:40.239966299 +0300
> @@ -1062,7 +1062,7 @@
>  master-bin.000001	#	Write_rows	#	#	table_id: # flags: STMT_END_F
>  master-bin.000001	#	Xid	#	#	COMMIT /* XID */
>  master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t1` /* generated by server */
> -master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2` /* generated by server */
> +master-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2`
>  # SHOW BINLOG EVENTS ON 2nd CONNECTION TO MASTE
>  set @@session.binlog_rows_query_log_events= off;
>  # Test statements from the 2nd connection have no comment event
> @@ -1146,7 +1146,7 @@
>  slave-bin.000001	#	Write_rows	#	#	table_id: # flags: STMT_END_F
>  slave-bin.000001	#	Xid	#	#	COMMIT /* XID */
>  slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t1` /* generated by server */
> -slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2` /* generated by server */
> +slave-bin.000001	#	Query	#	#	use `test`; DROP TABLE `t2`
> ---
> ```
> 
> Please fix all test failures. After ensuring all failures are existing failures, please re-request for review.

@nitendra-bhosle also regarding the failure, why only t2 is affected and not t1 ?

nitendra-bhosle commented:
> @nitendra-bhosle also regarding the failure, why only t2 is affected and not t1 ?

Let me check

nitendra-bhosle commented:
> > @nitendra-bhosle also regarding the failure, why only t2 is affected and not t1 ?
> 
> Let me check

Table t1 is created using engine = myisam 
create table t1(a int, b int) engine= myisam;

While table t2 is created using engine = innodb
create table t2(a int, b int) engine=innodb;

maybe this is leading to this behaviour @venkatesh-prasad-v  
What's your opinion on this?

nitendra-bhosle commented:
Jenkins test runs -- 
Branch 8.0-PS-7526 - https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1031/
Branch 8.0 - https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1029/

Failing test are not looking related to changes made by PS-7526

These are tests which are passing on retry , but showing as failed --

MySQL.rpl_gtid.rpl_mts_slave_preserve_commit_order_empty_gtid_nobinlog

satya-bodapati commented on mysql-test/extra/binlog_tests/binlog_drop_tbl_with_comment.test:
```diff
@@ -0,0 +1,56 @@
+
+SET binlog_ddl_skip_rewrite = OFF;
+
+/* comment1 */ CREATE TABLE t (a VARCHAR(100));
+/* comment1 */ DROP TABLE t;
```
While analyzing test results, I just realized, we didn't add a test for partitioned table. and also drop multi tables of partition type within a version comment marker

Can you please add these?

Test results look good now

nitendra-bhosle commented:
Jenkins run after making partition table drop statements  tests  : https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1033/

satya-bodapati reviewed (commented):
LGTM! @venkatesh-prasad-v is it OK to be squashed?

venkatesh-prasad-v commented on mysql-test/extra/binlog_tests/binlog_drop_tbl_with_comment.test:
```diff
@@ -21,8 +36,26 @@ SET binlog_ddl_skip_rewrite = ON;
+PARTITION BY KEY (a)
+PARTITIONS 2;
+
+--error ER_DROP_MULTI_TABLE
+/* comment at start */ DROP TABLE t1 /*!80017 ,t2*/ /* comment at end */;
```
Can you also add a scenario to test a single table DROP TABLE query with versioned comment?

`/* comment at start */ DROP TABLE t1 /*!80017 */ /* comment at end */;`

venkatesh-prasad-v commented:
Looks good. Please squash all the commits and update the description section with the updated commit message.
>Adde new option binlog_ddl_query_log_events , which is
by default OFF, and could be enabled by 'SET binlog_ddl_query_log_events = ON'
If enabled single table drop ddl statements will be logged as it is
without striping of comments and adding quotes.
Generate error for multi table drop table statements

As of now, the description still has the old variable name.

nitendra-bhosle commented:
> Looks good. Please squash all the commits and update the description section with the updated commit message.
> 
> > Adde new option binlog_ddl_query_log_events , which is
> > by default OFF, and could be enabled by 'SET binlog_ddl_query_log_events = ON'
> > If enabled single table drop ddl statements will be logged as it is
> > without striping of comments and adding quotes.
> > Generate error for multi table drop table statements
> 
> As of now, the description still has the old variable name.

Changed description.

nitendra-bhosle commented:
> LGTM! Good work on this. It seemed very simple but it was tricky

Yeah right .Thanks for reviewing and suggesting changes ,which are helpful.

venkatesh-prasad-v reviewed (approved):
Thank you very much for working on this issue and for diligently addressing all the review comments. The patch looks great and is approved from my side.

Regards,
